File: languages/csharp/AppEncryption/AppEncryption.IntegrationTests/Regression/AppEncryptionParameterizedTest.cs
using System;
using System.Collections;
using System.Collections.Generic;
using App.Metrics;
using GoDaddy.AppServices.AppEncryption.IntegrationTests.Regression;
using GoDaddy.Asherah.AppEncryption.Envelope;
using GoDaddy.Asherah.AppEncryption.IntegrationTests.TestHelpers;
using GoDaddy.Asherah.AppEncryption.IntegrationTests.Utils;
using GoDaddy.Asherah.AppEncryption.KeyManagement;
using GoDaddy.Asherah.AppEncryption.Persistence;
using GoDaddy.Asherah.AppEncryption.Util;
using GoDaddy.Asherah.Crypto;
using GoDaddy.Asherah.Crypto.Engine.BouncyCastle;
using GoDaddy.Asherah.Crypto.Keys;
using Moq;
using Newtonsoft.Json.Linq;
using Xunit;
using static GoDaddy.Asherah.AppEncryption.IntegrationTests.TestHelpers.Constants;

namespace GoDaddy.Asherah.AppEncryption.IntegrationTests.Regression
{
    public class AppEncryptionParameterizedTest
    {
        private readonly JObject payload;

        public AppEncryptionParameterizedTest()
        {
            payload = PayloadGenerator.CreateDefaultRandomJsonPayload();
        }

        [Theory]
        [ClassData(typeof(AppEncryptionParameterizedTestData))]
        public void ParameterizedTests(
            IEnvelopeEncryption<byte[]> envelopeEncryptionJson,
            Mock<MemoryPersistenceImpl<JObject>> metastorePersistence,
            KeyState cacheIK,
            KeyState metaIK,
            KeyState cacheSK,
            KeyState metaSK,
            AppEncryptionPartition appEncryptionPartition)
        {
            using (AppEncryption<JObject, byte[]> appEncryptionJsonImpl =
                new AppEncryptionJsonImpl<byte[]>(envelopeEncryptionJson))
            {
                EncryptMetastoreInteractions encryptMetastoreInteractions =
                    new EncryptMetastoreInteractions(cacheIK, metaIK, cacheSK, metaSK);
                DecryptMetastoreInteractions decryptMetastoreInteractions =
                    new DecryptMetastoreInteractions(cacheIK, cacheSK);

                // encrypt with library object(appEncryptionJsonImpl)
                byte[] encryptedPayload = appEncryptionJsonImpl.Encrypt(payload);

                Assert.NotNull(encryptedPayload);
                VerifyEncryptFlow(metastorePersistence, encryptMetastoreInteractions, appEncryptionPartition);

                metastorePersistence.Reset();
                JObject decryptedPayload = appEncryptionJsonImpl.Decrypt(encryptedPayload);

                VerifyDecryptFlow(metastorePersistence, decryptMetastoreInteractions, appEncryptionPartition);
                Assert.True(JToken.DeepEquals(payload, decryptedPayload));
            }
        }

        private void VerifyDecryptFlow(
            Mock<MemoryPersistenceImpl<JObject>> metastorePersistence,
            DecryptMetastoreInteractions metastoreInteractions,
            AppEncryptionPartition appEncryptionPartition)
        {
            // If IK is loaded from metastore
            if (metastoreInteractions.ShouldLoadIK())
            {
                metastorePersistence.Verify(
                    x => x.Load(appEncryptionPartition.IntermediateKeyId, It.IsAny<DateTimeOffset>()), Times.Once);
            }

            // If SK is loaded from metastore
            if (metastoreInteractions.ShouldLoadSK())
            {
                metastorePersistence.Verify(
                    x => x.Load(appEncryptionPartition.SystemKeyId, It.IsAny<DateTimeOffset>()),
                    Times.Once);
            }
        }

        private void VerifyEncryptFlow(
            Mock<MemoryPersistenceImpl<JObject>> metastorePersistence,
            EncryptMetastoreInteractions metastoreInteractions,
            AppEncryptionPartition appEncryptionPartition)
        {
            // If IK is stored to metastore
            if (metastoreInteractions.ShouldStoreIK())
            {
                metastorePersistence.Verify(
                    x => x.Store(appEncryptionPartition.IntermediateKeyId, It.IsAny<DateTimeOffset>(), It.IsAny<JObject>()),
                    Times.Once);
            }

            // If SK is stored to metastore
            if (metastoreInteractions.ShouldStoreSK())
            {
                metastorePersistence.Verify(
                    x => x.Store(appEncryptionPartition.SystemKeyId, It.IsAny<DateTimeOffset>(), It.IsAny<JObject>()),
                    Times.Once);
            }

            // If neither IK nor SK is stored
            if (!metastoreInteractions.ShouldStoreIK() && !metastoreInteractions.ShouldStoreSK())
            {
                metastorePersistence.Verify(
                    x => x.Store(It.IsAny<string>(), It.IsAny<DateTimeOffset>(), It.IsAny<JObject>()),
                    Times.Never);
            }

            // NOTE: We do not read IK from the metastore in case of Encrypt
            // If SK is loaded from metastore
            if (metastoreInteractions.ShouldLoadSK())
            {
                metastorePersistence.Verify(
                    x => x.Load(appEncryptionPartition.SystemKeyId, It.IsAny<DateTimeOffset>()),
                    Times.Once);
            }
            else
            {
                metastorePersistence.Verify(
                    x => x.Load(It.IsAny<string>(), It.IsAny<DateTimeOffset>()),
                    Times.Never);
            }

            // If latest IK is loaded from metastore
            if (metastoreInteractions.ShouldLoadLatestIK())
            {
                metastorePersistence.Verify(
                    x => x.LoadLatestValue(appEncryptionPartition.IntermediateKeyId),
                    Times.Once);
            }

            // If latest SK is loaded from metastore
            if (metastoreInteractions.ShouldLoadLatestSK())
            {
                metastorePersistence.Verify(
                    x => x.LoadLatestValue(appEncryptionPartition.SystemKeyId),
                    Times.Once);
            }

            // If neither latest IK or SK is loaded from metastore
            if (!metastoreInteractions.ShouldLoadLatestSK() && !metastoreInteractions.ShouldLoadLatestIK())
            {
                metastorePersistence.Verify(
                    x => x.LoadLatestValue(It.IsAny<string>()),
                    Times.Never);
            }
        }

        private class AppEncryptionParameterizedTestData : IEnumerable<object[]>
        {
            private static readonly Random Random = new Random();
            private readonly ConfigFixture configFixture;

            public AppEncryptionParameterizedTestData()
            {
                configFixture = new ConfigFixture();
            }

            public IEnumerator<object[]> GetEnumerator()
            {
                foreach (KeyState cacheIK in Enum.GetValues(typeof(KeyState)))
                {
                    foreach (KeyState metaIK in Enum.GetValues(typeof(KeyState)))
                    {
                        foreach (KeyState cacheSK in Enum.GetValues(typeof(KeyState)))
                        {
                            foreach (KeyState metaSK in Enum.GetValues(typeof(KeyState)))
                            {
                                // TODO Add CryptoPolicy.KeyRotationStrategy loop and update expect/verify logic accordingly
                                yield return GenerateMocks(cacheIK, metaIK, cacheSK, metaSK);
                            }
                        }
                    }
                }
            }

            IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();

            private object[] GenerateMocks(KeyState cacheIK, KeyState metaIK, KeyState cacheSK, KeyState metaSK)
            {
                AppEncryptionPartition appEncryptionPartition = new AppEncryptionPartition(
                    cacheIK + "CacheIK_" + metaIK + "MetaIK_" + DateTimeUtils.GetCurrentTimeAsUtcIsoDateTimeOffset() +
                    "_" + Random.Next(),
                    cacheSK + "CacheSK_" + metaSK + "MetaSK_" + DateTimeUtils.GetCurrentTimeAsUtcIsoDateTimeOffset() + "_" + Random.Next(),
                    DefaultProductId);

                KeyManagementService kms = configFixture.KeyManagementService;

                CryptoKeyHolder cryptoKeyHolder = CryptoKeyHolder.GenerateIKSK();

                Mock<MemoryPersistenceImpl<JObject>> metastorePersistence = MetastoreMock.CreateMetastoreMock(
                    appEncryptionPartition, kms, metaIK, metaSK, cryptoKeyHolder);

                CacheMock cacheMock = CacheMock.CreateCacheMock(cacheIK, cacheSK, cryptoKeyHolder);

                // Mimics (mostly) the old TimeBasedCryptoPolicyImpl settings
                CryptoPolicy cryptoPolicy = BasicExpiringCryptoPolicy.NewBuilder()
                    .WithKeyExpirationDays(KeyExpiryDays)
                    .WithRevokeCheckMinutes(int.MaxValue)
                    .WithCanCacheIntermediateKeys(false)
                    .WithCanCacheSystemKeys(false)
                    .Build();

                SecureCryptoKeyDictionary<DateTimeOffset> intermediateKeyCache = cacheMock.IntermediateKeyCache;
                SecureCryptoKeyDictionary<DateTimeOffset> systemKeyCache = cacheMock.SystemKeyCache;

                EnvelopeEncryptionJsonImpl envelopeEncryptionJson = new EnvelopeEncryptionJsonImpl(
                    appEncryptionPartition,
                    metastorePersistence.Object,
                    systemKeyCache,
                    new FakeSecureCryptoKeyDictionaryFactory<DateTimeOffset>(intermediateKeyCache),
                    new BouncyAes256GcmCrypto(),
                    cryptoPolicy,
                    kms);

                IEnvelopeEncryption<byte[]> envelopeEncryptionByteImpl =
                    new EnvelopeEncryptionBytesImpl(envelopeEncryptionJson);

                // Need to manually set a no-op metrics instance
                IMetrics metrics = new MetricsBuilder()
                    .Configuration.Configure(options => options.Enabled = false)
                    .Build();
                MetricsUtil.SetMetricsInstance(metrics);

                return new object[]
                {
                    envelopeEncryptionByteImpl, metastorePersistence, cacheIK, metaIK, cacheSK, metaSK,
                    appEncryptionPartition
                };
            }
        }
    }
}

File: tests/java/test-app/src/main/java/com/godaddy/asherah/testapp/regression/AppEncryptionParameterizedTest.java
package com.godaddy.asherah.testapp.regression;

import com.godaddy.asherah.appencryption.AppEncryption;
import com.godaddy.asherah.appencryption.AppEncryptionJsonImpl;
import com.godaddy.asherah.appencryption.AppEncryptionPartition;
import com.godaddy.asherah.appencryption.envelope.EnvelopeEncryption;
import com.godaddy.asherah.appencryption.envelope.EnvelopeEncryptionBytesImpl;
import com.godaddy.asherah.appencryption.envelope.EnvelopeEncryptionJsonImpl;
import com.godaddy.asherah.appencryption.keymanagement.KeyManagementService;
import com.godaddy.asherah.appencryption.persistence.MetastorePersistence;
import com.godaddy.asherah.crypto.BasicExpiringCryptoPolicy;
import com.godaddy.asherah.crypto.CryptoPolicy;
import com.godaddy.asherah.crypto.engine.bouncycastle.BouncyAes256GcmCrypto;
import com.godaddy.asherah.crypto.keys.SecureCryptoKeyMap;
import com.godaddy.asherah.testapp.TestSetup;
import com.godaddy.asherah.testapp.testhelpers.KeyState;
import com.godaddy.asherah.testapp.utils.DateTimeUtils;
import com.godaddy.asherah.testapp.utils.PayloadGenerator;

import org.json.JSONObject;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.extension.ExtendWith;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;
import org.mockito.junit.jupiter.MockitoExtension;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;
import java.util.stream.Stream;

import static com.godaddy.asherah.testapp.testhelpers.Constants.*;
import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
class AppEncryptionParameterizedTest {
  private static final Logger LOG = LoggerFactory.getLogger(AppEncryptionParameterizedTest.class);
  private static final Random RANDOM = new Random();

  private JSONObject payload;

  @BeforeEach
  public void setupTest() {
    payload = PayloadGenerator.createDefaultRandomJsonPayload();
  }

  @SuppressWarnings("unchecked")
  @DisplayName("App Encryption scenarios")
  @ParameterizedTest(name = "{index} => {2}CacheIK{4}MetastoreIK{3}CacheSK{5}MetastoreSK")
  @MethodSource("generateScenarios")
  void parameterizedTests(
      final EnvelopeEncryption<byte[]> envelopeEncryptionJson,
      final MetastorePersistence<JSONObject> metastorePersistence,
      final KeyState cacheIK, final KeyState metaIK,
      final KeyState cacheSK, final KeyState metaSK,
      final AppEncryptionPartition appEncryptionPartition) {

    try (AppEncryption<JSONObject, byte[]> appEncryptionJsonImpl = new AppEncryptionJsonImpl<>(envelopeEncryptionJson)) {

      EncryptMetastoreInteractions encryptMetastoreInteractions =
          new EncryptMetastoreInteractions(cacheIK, metaIK, cacheSK, metaSK);
      DecryptMetastoreInteractions decryptMetastoreInteractions =
          new DecryptMetastoreInteractions(cacheIK, cacheSK);

      //encrypt with library object(appEncryptionJsonImpl)
      byte[] encryptedPayload = appEncryptionJsonImpl.encrypt(payload);

      assertNotNull(encryptedPayload);
      verifyEncryptFlow(metastorePersistence, encryptMetastoreInteractions, appEncryptionPartition);

      reset(metastorePersistence);
      JSONObject decryptedPayload = appEncryptionJsonImpl.decrypt(encryptedPayload);

      verifyDecryptFlow(metastorePersistence, decryptMetastoreInteractions, appEncryptionPartition);
      assertTrue(payload.similar(decryptedPayload));
    }
  }

  private void verifyEncryptFlow(final MetastorePersistence<JSONObject> metastorePersistence,
      final EncryptMetastoreInteractions metastoreInteractions, final AppEncryptionPartition appEncryptionPartition) {

    // If IK is stored to metastore
    if (metastoreInteractions.shouldStoreIK()) {
      verify(metastorePersistence)
          .store(eq(appEncryptionPartition.getIntermediateKeyId()), any(Instant.class), any(JSONObject.class));
    }
    // If SK is stored to metastore
    if (metastoreInteractions.shouldStoreSK()) {
      verify(metastorePersistence)
          .store(eq(appEncryptionPartition.getSystemKeyId()), any(Instant.class), any(JSONObject.class));
    }
    // If neither IK nor SK is stored
    if (!metastoreInteractions.shouldStoreIK() && !metastoreInteractions.shouldStoreSK()) {
      verify(metastorePersistence,
          never()).store(any(String.class), any(Instant.class), any(JSONObject.class));
    }

    // NOTE: We do not read IK from the metastore in case of Encrypt
    // If SK is loaded from metastore
    if (metastoreInteractions.shouldLoadSK()) {
      verify(metastorePersistence)
          .load(eq(appEncryptionPartition.getSystemKeyId()), any(Instant.class));
    }
    else {
      verify(metastorePersistence,
          never()).load(anyString(), any(Instant.class));
    }

    // If latest IK is loaded from metastore
    if (metastoreInteractions.shouldLoadLatestIK()) {
      verify(metastorePersistence)
          .loadLatestValue(eq(appEncryptionPartition.getIntermediateKeyId()));
    }
    // If latest SK is loaded from metastore
    if (metastoreInteractions.shouldLoadLatestSK()) {
      verify(metastorePersistence)
          .loadLatestValue(eq(appEncryptionPartition.getSystemKeyId()));
    }
    // If neither latest IK or SK is loaded from metastore
    if (!metastoreInteractions.shouldLoadLatestSK() && !metastoreInteractions.shouldLoadLatestIK()) {
      verify(metastorePersistence,
          never()).loadLatestValue(any(String.class));
    }
  }

  private void verifyDecryptFlow(final MetastorePersistence<JSONObject> metastorePersistence,
      final DecryptMetastoreInteractions metastoreInteractions, final AppEncryptionPartition appEncryptionPartition) {

    // If IK is loaded from metastore
    if (metastoreInteractions.shouldLoadIK()) {
      verify(metastorePersistence)
          .load(eq(appEncryptionPartition.getIntermediateKeyId()), any(Instant.class));
    }

    // If SK is loaded from metastore
    if (metastoreInteractions.shouldLoadSK()) {
      verify(metastorePersistence)
          .load(eq(appEncryptionPartition.getSystemKeyId()), any(Instant.class));
    }
  }

  private static Stream<Arguments> generateScenarios() {
    List<Arguments> listOfScenarios = new ArrayList<>();
    for (KeyState cacheIK : KeyState.values()) {
      for (KeyState metaIK : KeyState.values()) {
        for (KeyState cacheSK : KeyState.values()) {
          for (KeyState metaSK : KeyState.values()) {
            // TODO Add CryptoPolicy.KeyRotationStrategy loop and update expect/verify logic accordingly
            Arguments arguments = generateMocks(cacheIK, metaIK, cacheSK, metaSK);
            listOfScenarios.add(arguments);
          }
        }
      }
    }

    return listOfScenarios.stream();
  }

  private static Arguments generateMocks(final KeyState cacheIK, final KeyState metaIK,
                                         final KeyState cacheSK, final KeyState metaSK) {

    AppEncryptionPartition appEncryptionPartition = new AppEncryptionPartition(
        cacheIK.toString() + "CacheIK_" + metaIK.toString() + "MetaIK_" + DateTimeUtils.getCurrentTimeAsUtcIsoOffsetDateTime() +
            "_" + RANDOM.nextInt(),
        cacheSK.toString() + "CacheSK_" + metaSK.toString() + "MetaSK_" + DateTimeUtils.getCurrentTimeAsUtcIsoOffsetDateTime() +
            "_" + RANDOM.nextInt(),
        DEFAULT_PRODUCT_ID);

    KeyManagementService kms = TestSetup.getKeyManagementService();

    CryptoKeyHolder cryptoKeyHolder = CryptoKeyHolder.generateIKSK();

    MetastorePersistence<JSONObject> metastorePersistence = MetastoreMock.createMetastoreMock(appEncryptionPartition, kms,
        TestSetup.getMetastorePersistence(), metaIK, metaSK, cryptoKeyHolder);

    CacheMock cacheMock = CacheMock.createCacheMock(cacheIK, cacheSK, cryptoKeyHolder);

    // Mimics (mostly) the old TimeBasedCryptoPolicyImpl settings
    CryptoPolicy cryptoPolicy = BasicExpiringCryptoPolicy.newBuilder()
        .withKeyExpirationDays(KEY_EXPIRY_DAYS)
        .withRevokeCheckMinutes(Integer.MAX_VALUE)
        .withCanCacheIntermediateKeys(false)
        .withCanCacheSystemKeys(false)
        .build();

    SecureCryptoKeyMap<Instant> intermediateKeyCache = cacheMock.getIntermediateKeyCache();
    SecureCryptoKeyMap<Instant> systemKeyCache = cacheMock.getSystemKeyCache();

    EnvelopeEncryptionJsonImpl envelopeEncryptionJson = new EnvelopeEncryptionJsonImpl(
        appEncryptionPartition, metastorePersistence, systemKeyCache,
        new FakeSecureCryptoKeyMapFactory<>(intermediateKeyCache), new BouncyAes256GcmCrypto(),
        cryptoPolicy, kms
    );

    EnvelopeEncryption<byte[]> envelopeEncryptionByteImpl = new EnvelopeEncryptionBytesImpl(envelopeEncryptionJson);

    return Arguments.of(envelopeEncryptionByteImpl, metastorePersistence,
        cacheIK, metaIK, cacheSK, metaSK, appEncryptionPartition);
  }
}
